### ISSUE TITLE
Parallel execution of async rules

### ISSUE THREAD

**Issue by ccounselman** [FluentValidation/FluentValidation#2275]
### Is your feature request related to a problem? Please describe.

I'm working on a microservice project that requires validations to check multiple external services.
As a trivial example, consider an AddOrder endpoint of an Order service that has to verify incoming requests:
- Call Customer service to validate Order.CustomerId, and
- Call ProductCatalog service to validate Order.ProductId.

Given a validator that performs multiple, independent, asynchronous validations, the [current implementation](https://github.com/FluentValidation/FluentValidation/blob/main/src/FluentValidation/AbstractValidator.cs#L246) executes and awaits each rule in sequential order, i.e. each rule needs to complete before the next one starts.

```csharp
for (int i = 0; i < count; i++) {
    cancellation.ThrowIfCancellationRequested();
    var totalFailures = context.Failures.Count;

    await Rules[i].ValidateAsync(context, useAsync, cancellation);

    if (ClassLevelCascadeMode == CascadeMode.Stop && result.Errors.Count > totalFailures) {
        // Bail out if we're "failing-fast". Check to see if the number of failures
	// has been increased by this rule (which could've generated 1 or more failures).
	break;
    }
}
```

If there are 5 validation rules that each spend 1 second waiting on an I/O operation... the total validation requires 5 seconds to complete.

If the rules could be executed in parallel, then performance could be improved -- the 5 x 1 second validations could all execute in 1 second (plus some overhead).

### Describe the solution you'd like

Support overriding the default behavior of AbstractValidator via an Enum property (similar to CascadeMode):
```csharp
public enum AsyncExecutionMode {
    Sequential = 0,
    Multithreaded,
    Parallel
}

public abstract class AbstractValidator<T> {
    public virtual AsyncExecutionMode AsyncExecutionMode { get; set; }
   ...

    private async ValueTask<ValidationResult> ValidateInternalAsync(ValidationContext<T> context, bool useAsync, CancellationToken cancellation) {
        ...

        if (!useAsync || AsyncExecutionMode == AsyncExecutionMode.Sequential) {
            await RunRulesInSequenceAsync(context, result, useAsync, cancellation);
	}
	else if (AsyncExecutionMode == AsyncExecutionMode.Multithreaded) {
	    await RunRulesMultithreadedAsync(context, useAsync, cancellation);
	}
	else if (AsyncExecutionMode == AsyncExecutionMode.Parallel) {
	    await RunRulesInParallelAsync(context, useAsync, cancellation);
	}
	else {
	    throw new InvalidOperationException($"Unhandled asynchronous execution mode: {AsyncExecutionMode}");
	}

        ...
    }
}
```

so that asynchronous calls could be executed in parallel:

```csharp
// Begin executing all rules.
var tasks = new ValueTask[count];
for (int i = 0; i < count; i++) {
    cancellation.ThrowIfCancellationRequested();
    tasks[i] = Rules[i].ValidateAsync(context, useAsync, cancellation);
}

// Wait for all rules to complete.
for (int i = 0; i < count; i++) {
    cancellation.ThrowIfCancellationRequested();
    await tasks[i];
}
```

or possibly even multi-threaded:

```csharp
// Execute each rule on a separate thread.
var tasks = new Task[count];
for (int i = 0; i < count; i++) {
    cancellation.ThrowIfCancellationRequested();
    tasks[i] = Task.Run(() => Rules[i].ValidateAsync(context, useAsync, cancellation).AsTask(), cancellation);
}

// Wait for all rule to complete.
await Task.WhenAll(tasks).ConfigureAwait(false);
```

#### Additional considerations
- If rules are executing in parallel, then CascadeMode might need to be ignored -- there's really no "first".
- Existing behavior (sequential awaits) should remain default to remain backwards compatible.
  - Need to opt-in to parallel executions by changing the mode property.

### Describe alternatives you've considered

Tried creating a new subclass (AbstractParallelValidator) and overriding `ValidateAsync`, but certain parts of the framework (e.g. IValidationRuleInternal) are internal.

### Additional Context

[More detailed code here](https://github.com/ccounselman/FluentValidation/pull/1)

**Comment by JeremySkinner** [comment]
Hi, thanks for the suggestion but this isn't something we'll be adding to the project. Parallel rule execution of rules is outside of the scope of FluentValidation, and is not the intention behind async rules. FluentValidation isn't the right tool to be using if you're looking for parallel execution.
